Source/Utility/UDim4.lua

```diff
@@ -1,34 +1,124 @@
-local UDim4 = {}
-UDim4.__index = UDim4
+-- @author MemoryPenguin
+-- @author Narrev
+-- UDim4 Object
 
-function UDim4.new(x, y)
-	assert(x and y, "Must supply x and y values.")
-	
-	local self = setmetatable({
-		X = x;
-		Y = y;
-	}, UDim4)
-	
-	return self
+local newVector2 = Vector2.new
+local newUDim2 = UDim2.new
+local IsA = game.IsA
+local SizePos = {
+	Size = "AbsoluteSize";
+	Position = "AbsolutePosition";
+}
+
+local function Connect(Event, func)
+	Connect = Event.Connect
+	Connect(Event, func)
+end
+
+local function Disconnect(Event)
+	Disconnect = Event.Disconnect
+	Disconnect(Event)
 end
 
-function UDim4.FromCoordinates(...)
-	return UDim4.new(
-		UDim2.new(select(1, ...)),
-		UDim2.new(select(5, ...))
+local function ToUDim2(self, GuiObject)
+	if not GuiObject.Parent or not IsA(GuiObject.Parent, "GuiBase2d") then
+		return newVector2(0, 0)
+	end
+
+	local ParentSize = GuiObject.Parent.AbsoluteSize
+	local X, Y, XX, XY, YX, YY = ParentSize.X, ParentSize.Y, self.XX, self.XY, self.YX, self.YY
+	return newUDim2(
+		0, XX.Offset + XY.Offset + XX.Scale*X + XY.Scale*Y,
+		0, YY.Offset + YX.Offset + YY.Scale*Y + YX.Scale*X
 	)
 end
 
-function UDim4:ToVector2(guiObject)
-	if not guiObject.Parent or not guiObject.Parent:IsA("GuiBase2d") then
-		return Vector2.new(0, 0)
+local function Render(self)
+	--- Renders position and size
+
+	local Object = self.Object
+	Object.Position = ToUDim2(self.AbsolutePosition, Object)
+	Object.Size = ToUDim2(self.AbsoluteSize, Object)
+end
+
+local function AttachToParent(self)
+	--- Attaches to object's parent and intercepts resize events there
+
+	local Parent = self.Object.Parent
+
+	if self._parentConnection then
+		Disconnect(self._parentConnection)
+	end
+
+	if Parent then
+		self._parentConnection = Connect(Parent.Changed, function(Property)
+			if Property == "AbsoluteSize" or Property == "AbsolutePosition" then
+				Render(self)
+			end
+		end)
+	end
+
+	Render(self)
+	return self
+end
+
+function newUDim4(X, Y, ...)
+	if ... then
+		return setmetatable({X = newUDim2(X, Y, ...); Y = newUDim2(select(3, ...))}, UDim4)
+	else
+		return setmetatable({X = X or newUDim2(); Y = Y or newUDim2()}, UDim4)
+	end
+end
+
+-- UDim4Render Class
+local UDim4Render = {Render = Render}
+
+function UDim4Render:__index(i)
+	return UDim4Render[i] or self[SizePos[i]]
+end
+
+function UDim4Render:__newindex(i, v)
+	self[SizePos[i]] = v -- Only allows you to change Position and Size
```

Comment:
Don't rely on Lua builtin behavior to restrict access like this; use an `if` statement to provide a clearer error message than `table index is nil`.